File: smartmoneyserver/src/main/java/fpt/aptech/server/dto/budget/BudgetResponse.java
package fpt.aptech.server.dto.budget;

import fpt.aptech.server.dto.category.CategoryResponse;
import fpt.aptech.server.enums.budget.BudgetType;
import lombok.Builder;
import lombok.Getter;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.List;

@Getter
@Builder
public class BudgetResponse {

    // ── Thông tin cơ bản ──────────────────────────────────────────────────────
    private Integer id;
    private BigDecimal amount;          // Hạn mức ngân sách
    private LocalDate beginDate;
    private LocalDate endDate;
    private Integer walletId;           // null = tất cả ví
    private String walletName;          // null = "Tổng cộng"
    private Boolean allCategories;
    private Boolean repeating;

    private    Boolean exceeded;// đã vượt
    private    Boolean warning; // sắp vượt
    private    BigDecimal progress; // % tiến độ (0 → 1)

    private List<CategoryResponse> categories; // rỗng nếu allCategories=true

    // ── Icon danh mục chính (Flutter hiển thị) ────────────────────────────────
    private Integer primaryCategoryId;          // 1. ID danh mục chính (null nếu allCategories=true)
    private String primaryCategoryIconUrl;      // 2. Icon URL danh mục chính (null nếu allCategories=true)

    // ── Trạng thái ───────────────────────────────────────────────────────────
    private boolean expired;            // endDate < today

    // ── Thực chi ─────────────────────────────────────────────────────────────
    private BigDecimal spentAmount;     // Đã chi
    private BigDecimal remainingAmount; // Còn lại (amount - spentAmount)

    // ── Chỉ số dự đoán ───────────────────────────────────────────────────────
    private BigDecimal dailyShouldSpend;  // Số tiền nên chi mỗi ngày để không vượt ngân sách = remainingAmount / daysLeft
    private BigDecimal dailyActualSpend;  // Trung bình chi mỗi ngày thực tế = spentAmount / daysElapsed (từ ngày bắt đầu đến hôm nay)
    private BigDecimal projectedSpend;    // Dự kiến tổng chi nếu tiếp tục mức chi hiện tại = dailyActual * totalDays

    // ── Đề xuất dựa trên lịch sử 3 tháng ───────────────────────────────────────
    private BigDecimal suggestedAmount;       // Ngân sách đề xuất dựa trên lịch sử chi tiêu 3 tháng gần nhất
    private BigDecimal suggestedDailySpend;   // Trung bình chi mỗi ngày từ lịch sử 3 tháng
    private BigDecimal suggestedWeeklySpend;  // Ngân sách đề xuất hàng tuần (dailyAverage * 7)
    private BigDecimal suggestedMonthlySpend; // Ngân sách đề xuất hàng tháng (dailyAverage * số ngày thực tế của tháng)
    private BigDecimal suggestedYearlySpend;  // Ngân sách đề xuất hàng năm (dailyAverage * số ngày thực tế của năm)
    private BigDecimal suggestedCustomSpend;  // Ngân sách đề xuất custom period (dailyAverage * số ngày custom)
    private BudgetType budgetType;            // Loại ngân sách (WEEKLY, MONTHLY, YEARLY, CUSTOM)
    private BigDecimal overBudgetAmount;       // Số tiền vượt ngân sách = max(0, spentAmount - amount)
}

File: APP MOBILE/smart_money/lib/modules/budget/models/budget_response.dart
import 'package:smart_money/modules/transaction/models/view/transaction_response.dart';

import '../../category/models/category_response.dart';
import '../enums/budget_type.dart';

class BudgetResponse {
  final int id;
  final double amount;
  final DateTime beginDate;
  final DateTime endDate;
  final int? walletId;
  final String? walletName;
  final bool allCategories;
  final bool repeating;
  final List<CategoryResponse> categories;
  final bool isOther;
  final BudgetType budgetType;

  final int? primaryCategoryId;
  final String? primaryCategoryIconUrl;

  final bool expired;

  final double spentAmount;
  final double remainingAmount;

  final double dailyShouldSpend;
  final double dailyActualSpend;
  final double projectedSpend;

  final double suggestedAmount;       // Ngân sách đề xuất dựa trên lịch sử 3 tháng
  final double suggestedDailySpend;   // Nên chi hàng ngày theo đề xuất
  final double? suggestedWeeklySpend;  // Nên chi hàng tuần theo đề xuất (chỉ có giá trị khi budgetType=WEEKLY)
  final double? suggestedMonthlySpend; // Nên chi hàng tháng theo đề xuất (chỉ có giá trị khi budgetType=MONTHLY)
  final double? suggestedYearlySpend;  // Nên chi hàng năm theo đề xuất (chỉ có giá trị khi budgetType=YEARLY)
  final double? suggestedCustomSpend;  // Nên chi theo custom period theo đề xuất (chỉ có giá trị khi budgetType=CUSTOM)
  final double overBudgetAmount;      // Số tiền vượt ngân sách = max(0, spent - amount)

  final bool exceeded;
  final bool warning;
  final double progress;

  const BudgetResponse({
    required this.id,
    required this.amount,
    required this.beginDate,
    required this.endDate,
    this.walletId,
    this.walletName,
    this.allCategories = false,
    this.repeating = false,
    this.categories = const [],
    this.primaryCategoryId,
    this.primaryCategoryIconUrl,
    this.expired = false,
    this.spentAmount = 0,
    this.remainingAmount = 0,
    this.dailyShouldSpend = 0,
    this.dailyActualSpend = 0,
    this.projectedSpend = 0,
    this.suggestedAmount = 0,
    this.suggestedDailySpend = 0,
    this.suggestedWeeklySpend,
    this.suggestedMonthlySpend,
    this.suggestedYearlySpend,
    this.suggestedCustomSpend,
    this.overBudgetAmount = 0,
    this.isOther = false,
    this.exceeded = false,
    this.warning = false,
    this.progress = 0,
    required this.budgetType,
  });

  /// Factory parse từ JSON
  factory BudgetResponse.fromJson(Map<String, dynamic> json) {
    final rawCategories = json['categories'];
    final List<CategoryResponse> safeCategories = (rawCategories is List)
        ? rawCategories
        .where((e) => e != null)
        .map((e) => CategoryResponse.fromJson(e as Map<String, dynamic>))
        .toList()
        : [];

    DateTime safeParseDate(String? value) {
      if (value == null || value.isEmpty) return DateTime.now();
      return DateTime.tryParse(value) ?? DateTime.now();
    }

    bool safeBool(dynamic v) => v == true;

    double safeDouble(dynamic v) {
      if (v == null) return 0;
      if (v is num) return v.toDouble();
      return double.tryParse(v.toString()) ?? 0;
    }

    return BudgetResponse(
      id: json['id'] ?? 0,
      amount: safeDouble(json['amount']),
      beginDate: safeParseDate(json['beginDate']),
      endDate: safeParseDate(json['endDate']),
      walletId: json['walletId'],
      walletName: json['walletName'],
      allCategories: safeBool(json['allCategories']),
      repeating: safeBool(json['repeating']),
      categories: safeCategories,
      primaryCategoryId: json['primaryCategoryId'],
      primaryCategoryIconUrl: json['primaryCategoryIconUrl'],
      expired: safeBool(json['expired']),
      spentAmount: safeDouble(json['spentAmount']),
      suggestedAmount: safeDouble(json['suggestedAmount']),
      suggestedDailySpend: safeDouble(json['suggestedDailySpend']),
      suggestedWeeklySpend: safeDouble(json['suggestedWeeklySpend']),
      suggestedMonthlySpend: safeDouble(json['suggestedMonthlySpend']),
      suggestedYearlySpend: safeDouble(json['suggestedYearlySpend']),
      suggestedCustomSpend: safeDouble(json['suggestedCustomSpend']),
      overBudgetAmount: safeDouble(json['overBudgetAmount']),
      remainingAmount: safeDouble(json['remainingAmount']),
      dailyShouldSpend: safeDouble(json['dailyShouldSpend']),
      dailyActualSpend: safeDouble(json['dailyActualSpend']),
      projectedSpend: safeDouble(json['projectedSpend']),
      exceeded: safeBool(json['exceeded']),
      warning: safeBool(json['warning']),
      progress: safeDouble(json['progress']),
      budgetType: BudgetTypeExtension.fromString(json['budgetType']),
      isOther: false,
    );
  }

  /// CopyWith duy nhất, đầy đủ tất cả trường
  BudgetResponse copyWith({
    double? amount,
    DateTime? beginDate,
    DateTime? endDate,
    List<CategoryResponse>? categories,
    bool? repeating,
    int? walletId,
    String? walletName,
    String? primaryCategoryIconUrl,
    bool? allCategories,
    bool? expired,
    double? spentAmount,
    double? remainingAmount,
    double? dailyShouldSpend,
    double? dailyActualSpend,
    double? projectedSpend,
    double? suggestedAmount,
    double? suggestedDailySpend,
    double? suggestedWeeklySpend,
    double? suggestedMonthlySpend,
    double? suggestedYearlySpend,
    double? suggestedCustomSpend,
    double? overBudgetAmount,
    bool? exceeded,
    bool? warning,
    double? progress,
    BudgetType? budgetType,
    bool? isOther,

  }) {
    return BudgetResponse(
      id: id,
      amount: amount ?? this.amount,
      beginDate: beginDate ?? this.beginDate,
      endDate: endDate ?? this.endDate,
      walletId: walletId ?? this.walletId,
      walletName: walletName ?? this.walletName,
      allCategories: allCategories ?? this.allCategories,
      repeating: repeating ?? this.repeating,
      categories: categories ?? this.categories,
      primaryCategoryId: primaryCategoryId,
      primaryCategoryIconUrl: primaryCategoryIconUrl ?? this.primaryCategoryIconUrl,
      expired: expired ?? this.expired,
      spentAmount: spentAmount ?? this.spentAmount,
      remainingAmount: remainingAmount ?? this.remainingAmount,
      dailyShouldSpend: dailyShouldSpend ?? this.dailyShouldSpend,
      dailyActualSpend: dailyActualSpend ?? this.dailyActualSpend,
      projectedSpend: projectedSpend ?? this.projectedSpend,
      suggestedAmount: suggestedAmount ?? this.suggestedAmount,
      suggestedDailySpend: suggestedDailySpend ?? this.suggestedDailySpend,
      suggestedWeeklySpend: suggestedWeeklySpend ?? this.suggestedWeeklySpend,
      suggestedMonthlySpend: suggestedMonthlySpend ?? this.suggestedMonthlySpend,
      suggestedYearlySpend: suggestedYearlySpend ?? this.suggestedYearlySpend,
      suggestedCustomSpend: suggestedCustomSpend ?? this.suggestedCustomSpend,
      overBudgetAmount: overBudgetAmount ?? this.overBudgetAmount,
      exceeded: exceeded ?? this.exceeded,
      warning: warning ?? this.warning,
      progress: progress ?? this.progress,
      budgetType: budgetType ?? this.budgetType,
      isOther: isOther ?? this.isOther,
    );
  }
}
